### DISCUSSION TITLE
Inconsistent mutation numbers between true trees and tsinferred trees

### DISCUSSION THREAD

**Discussion by Proteios1998** [tskit-dev/tsinfer#801] [Q&A]
Hi, I tried to use tsinfer to infer a tree sequence given the .samples file generated from the true simulated tree. I use `tsinfer.SampleData.from_tree_sequence(ts, path=out, num_flush_threads=10)` to get the .samples file. I used `tsinfer ls *.samples` to check the number of sites in the .samples file and the number is the same as the number of sites in the original tree sequence. However, after using tsinfer to get the inferred tree, I found the number of mutations in the inferred tree (581937) is different from that in the true tree (578231). Could you help explain why?

The information of the .samples file:
<img width="1335" alt="image" src="https://user-images.githubusercontent.com/44884282/219088679-4ad7aef4-4d07-4017-9257-3f3374c3c6e0.png">

The information of the inferred tree:
<img width="514" alt="image" src="https://user-images.githubusercontent.com/44884282/219088847-67a45719-ea0b-47c6-be57-c7d882bcb24d.png">

The information of the true tree:
<img width="479" alt="image" src="https://user-images.githubusercontent.com/44884282/219089247-08d6b5b8-1f92-41fa-b35d-ab2eeb1abb83.png">

**Comment by benjeffery** [comment]
One shouldn't expect that the inferred tree sequence will be the same as the true tree, only that the resulting genotypes of the samples will be the same. As such the number of mutations can differ, as if the topology of the tree is different a different set of mutations might be needed. There can also be back mutations depending on your tsinfer settings. See https://tskit.dev/tsinfer/docs/stable/inference.html#matching-ancestors-samples

**Reply by Proteios1998** [reply]
Thank you for your reply. But I am also wondering if the genotype in the .samples file already differs from the genotype of the true tree. I have done another inference before, using the following code to generate the .samples file:
```
with tsinfer.SampleData(
    path="simulation.samples", sequence_length=ts.sequence_length, num_flush_threads=2
) as sample_data:
    for var in ts.variants():
        sample_data.add_site(var.site.position, var.genotypes, var.alleles)
        progress.update()
    progress.close()
```

By doing this, the number of mutations in the inferred tree matches that in the true tree (with the same tsinfer settings). Is there a way to check the number of mutations in the .samples file?

**Reply by hyanwong** [reply]
The .samples file doesn't contain any mutations. It just describes the variation without postulating *how* it arose (i.e. without proposing mutations in the past).

If the "genotypes" of the samples differ from those in the resulting inferred tree sequence (other than where missing data has been imputed), that's definitely a bug. You should be able to check with the following code (which I am about to add to the tutorial):

```python
import numpy as np
for v_orig, v_inferred in zip(sample_data.variants(), inferred_ts.variants()):
    if any(
        np.array(v_orig.alleles)[v_orig.genotypes] !=
        np.array(v_inferred.alleles)[v_inferred.genotypes]
    ):
        raise ValueError("Genotypes in original dataset and inferred tree seq not equal")
print("Genotypes in original dataset and inferred tree sequence are the same")
```

**Comment by hyanwong** [comment]
Did you infer with a recombination rate or recombination map? If so, then `tsinfer` may have placed multiple mutations to explain some of the sites, rather than making a new tree. You can check which sites have multiple mutations using

```python
for s in inferred_ts.sites():
    if len(s.mutations) > 1:
        print(s)  # you might get a lot of sites printed here!
```

**Comment by hyanwong** [comment]
Did you ever manage fo figure out what was going on here, @Proteios1998 ?
